v2/internal/io/skip_writer_read_seeker.go

```diff
@@ -0,0 +1,108 @@
+package io
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+)
+
+// SkipWriterReaderSeeker wraps a factory producing a writer with a ReadSeeker implementation.
+// Note that Read and Seek are not thread-safe, they must not be called
+// concurrently.
+type SkipWriterReaderSeeker struct {
+	parentCtx context.Context
+	offset    uint64
+	size      uint64
+
+	cons          ReWriter
+	reader        *io.PipeReader
+	writeCancel   context.CancelFunc
+	writeComplete chan struct{}
+}
+
+// ReWriter is a function writing to an io.Writer from an offset.
+type ReWriter func(ctx context.Context, skip uint64, w io.Writer) (uint64, error)
+
+var _ io.ReadSeeker = (*SkipWriterReaderSeeker)(nil)
+
+// NewSkipWriterReaderSeeker creates an io.ReadSeeker around a ReWriter.
+func NewSkipWriterReaderSeeker(ctx context.Context, size uint64, cons ReWriter) *SkipWriterReaderSeeker {
```

Comment:
could we have a better name than `cons`? it's not obvious to me what this name is supposed to be (`consume`?), it's puzzling when we get to `c.cons()` and it seems like it should be more obvious what it's doing - should it be read as "start consuming now"?